lib/Sema/TypeCheckConcurrency.cpp

```diff
@@ -578,6 +587,61 @@ findMemberReference(Expr *expr) {
   return None;
 }
 
+static bool isAsyncCall(const ApplyExpr *call) {
+  // Effectively the same as doing a
+  // `dyn_cast_or_null<FunctionType>(call->getFn()->getType())`, check the
+  // result of that and then checking `isAsync` if it's defined.
+  Type funcTypeType = call->getFn()->getType();
+  if (!funcTypeType)
+    return false;
+  FunctionType *funcType = funcTypeType->getAs<FunctionType>();
+  return funcType && funcType->isAsync();
+}
+
+static llvm::SmallVector<InOutArgInfo, 3> extractAsyncInoutArgs(const ApplyExpr *call) {
+  // Is the function we're calling async? If not, it doesn't matter.
+  // Is the function we're calling a mutating method? (implicit inout of 'self')
+  // gather the inout parameters recursively
+
+  // 1.
+  // Is it async?
+  if (!isAsyncCall(call))
+    return {};
+
+  llvm::SmallVector<InOutArgInfo, 3> outputs;
+  // Collect all of the inout arguments from the arguments of an expression.
+  // This does not collect things that have been curried in
+  // isPartialApply should be true if the parameter is part of the partial
+  // application.
+  auto extractInoutArgs = [&outputs](const ApplyExpr *call,
+                                     bool isPartialApply=false) {
+    if (ParenExpr *arg = dyn_cast<ParenExpr>(call->getArg())) {
+      if (InOutExpr *inoutArg = dyn_cast<InOutExpr>(arg->getSubExpr()))
+        outputs.push_back({inoutArg, isPartialApply});
+    } else if (TupleExpr *argList = dyn_cast<TupleExpr>(call->getArg())) {
+      outputs.reserve(outputs.size() + argList->getNumElements());
+      for (Expr *arg : argList->getElements())
+        if (InOutExpr *inoutArg = dyn_cast<InOutExpr>(arg))
+          outputs.push_back({inoutArg, isPartialApply});
+    } else if (InOutExpr *arg = dyn_cast<InOutExpr>(call->getArg())) {
+      outputs.push_back({arg, isPartialApply});
+    }
+  };
+
+  // 2.
+  // Does the call have a nested partial application?
+  // If part of the partial application takes something inout, then we should
+  // track it. A struct-type property of an actor that calls a mutating func is
+  // implicitly passed inout to the mutating function.
+  if (ApplyExpr * subCall = dyn_cast<ApplyExpr>(call->getFn()))
```

Comment:
Nitpick: I would suggest being consistent with the `*` and either put it next to the type name, or the variable identifier, but not between the two like this. You seem to like the latter so I'd do that.